test/e2e/route_metrics_test.go

```diff
@@ -0,0 +1,393 @@
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+
+	// Add RouteSelector if testRS is set to true.
+	if testRS {
+		ic.Spec.RouteSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": correctLabel,
+			},
+		}
+	}
+
+	// Add NamespaceSelector if testNS is set to true.
+	if testNS {
+		ic.Spec.NamespaceSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": correctLabel,
+			},
+		}
+	}
+
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+
+	// Delete the IC if any error occurs.
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	// Store the start time to get the time taken to update metrics after the creation of the IC.
+	startTime := time.Now()
+
+	// Wait for metrics to be added and set to 0.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+
+	// Get the time taken for the metrics to be updated.
+	t.Logf("time taken for the metrics to be updated after the creation of the IC: %fs", time.Since(startTime).Seconds())
+
+	// Create a new namespace for the Route.
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: names.SimpleNameGenerator.GenerateName("test-e2e-metrics-"),
+		},
+	}
+
+	// Add Label to the Namespace if testRS is set to true.
+	if testNS {
+		ns.Labels = map[string]string{
+			"type": correctLabel,
+		}
+	}
+	if err := kclient.Create(context.TODO(), ns); err != nil {
+		t.Fatalf("failed to create namespace: %v", err)
+	}
+
+	// Delete the Namespace if any error occurs.
+	defer func() {
+		if err := kclient.Delete(context.TODO(), ns); err != nil {
+			t.Fatalf("failed to delete test namespace %s: %v", ns.Name, err)
+		}
+	}()
+
+	// Create a Route to be immediately admitted by this Ingress Controller.
+	routeFooLabelName := types.NamespacedName{Namespace: ns.Name, Name: routeNameStr}
+	routeFooLabel := newRouteWithLabel(routeFooLabelName, correctLabel)
+	if err := kclient.Create(context.TODO(), routeFooLabel); err != nil {
+		t.Fatalf("failed to create route: %v", err)
+	}
+
+	// Delete the Route if any error occurs.
+	defer func() {
+		if err := kclient.Delete(context.TODO(), routeFooLabel); err != nil {
+			if apierrors.IsNotFound(err) {
+				return
+			} else {
+				t.Fatalf("failed to delete route %s: %v", routeFooLabelName, err)
+			}
+		}
+	}()
+
+	// Wait for metrics to be updated to 1 as the Route will get admitted by the IC.
+	if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 1); err != nil {
+		t.Fatalf("failed to fetch expected metrics: %v", err)
+	}
+
+	if testNS {
+		// Update the NamespaceSelector of the Ingress Controller so that the Route gets un-admitted.
+		if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+			t.Fatalf("failed to get ingress resource: %v", err)
+		}
```

Comment:
If this Get is repeated for all cases, it can be moved out of the conditional case.

Reply:
The `Get` is required before every `Update` to get the latest version of the object. Tried moving it outside the conditional case and got the following error:
`failed to update ingresscontroller: Operation cannot be fulfilled on ingresscontrollers.operator.openshift.io "ic-rs-metrics-test": the object has been modified; please apply your changes to the latest version and try again`

Reply:
This is probably due to the fact that these tests run in parallel.

Reply:
I have run this individually as well and I still got the error. Also when the tests are run in parallel, the objects are specifically given different names so that there's no clash and the same object is not accessed in the different tests. So it is not a problem with the tests being run in parallel.

Reply:
I'd imagine it's specifically because of the `testNS=true` and `testRS=true` case in which testNS logic modifies the IC and testRS needs the up-to-date IC object. So this looks fine :+1: